### ISSUE TITLE
Exceptions inside a transaction are not shown

### ISSUE THREAD

**Issue by samphilipd** [elixir-ecto/ecto#2767]
If we raise an exception inside a transaction, the reported error is `DBConnection.ConnectionError`. There is no information about which line raised or what went wrong.

This makes it really difficult to debug the actual exception itself - is there some way of bubbling up the actual error that occurred, along with it's stacktrace.

**Comment by josevalim** [comment]
What Ecto version?

**Comment by josevalim** [comment]
We have an issue templates we ask developers to fill in when reporting a new issue, if you can do that, it would be very helpful.

**Comment by samphilipd** [comment]
### Environment

* Elixir version (elixir -v):

```
Erlang/OTP 21 [erts-10.0.7] [source] [64-bit] [smp:4:4] [ds:4:4:10] [async-threads:1] [hipe]

Elixir 1.7.3 (compiled with Erlang/OTP 21)
```

* Database and version (PostgreSQL 9.4, MongoDB 3.2, etc.):

Postgres 9.6

* Ecto version (mix deps):

```
* ecto 2.2.11 (Hex package) (mix)
  locked at 2.2.11 (ecto) 4bb8f117
```

* Database adapter and version (mix deps):
```
* postgrex 0.13.5 (Hex package) (mix)
  locked at 0.13.5 (postgrex) 3d931aba
```

* Operating system:

Mac OS X Mojave

### Steps to reproduce

```
# Example query from our codebase
import Ecto.Query

Repo.transaction(fn ->
  query =
    from(b in Buyer,
      order_by: [asc: :id]
    )

  query
  |> Repo.stream()
  |> Stream.each(fn buyer ->
    Repo.update!(Ecto.Changeset.change(buyer, %{close_io_lead_id: nil})) # Fails with not-null constraint
  end)
  |> Stream.run()
end)
```

### Current behavior

Running prints this:

```
[debug] QUERY OK db=0.8ms
begin []
[debug] QUERY OK source="buyers" db=7.1ms
SELECT b0."id", b0."close_io_lead_id", b0."name", b0."address", b0."lead_status", b0."notes", b0."inserted_at", b0."updated_at" FROM "buyers" AS b0 ORDER BY b0."id" []
[debug] QUERY OK source="buyers" db=7.5ms decode=2.5ms
SELECT b0."id", b0."close_io_lead_id", b0."name", b0."address", b0."lead_status", b0."notes", b0."inserted_at", b0."updated_at" FROM "buyers" AS b0 ORDER BY b0."id" []
[debug] QUERY ERROR db=17.1ms
UPDATE "buyers" SET "close_io_lead_id" = $1, "updated_at" = $2 WHERE "id" = $3 [nil, {{2018, 10, 23}, {11, 19, 15, 568676}}, 1]
[debug] QUERY OK db=3.5ms
rollback []
** (DBConnection.ConnectionError) transaction rolling back
    (db_connection) lib/db_connection.ex:1531: DBConnection.fetch_info/1
    (db_connection) lib/db_connection.ex:746: DBConnection.run/3
    (db_connection) lib/db_connection.ex:1509: DBConnection.deallocate/4
    (elixir) lib/stream.ex:1370: Stream.do_resource/5
    (elixir) lib/stream.ex:1553: Enumerable.Stream.do_each/4
    (elixir) lib/stream.ex:911: Stream.do_list_transform/7
    (elixir) lib/stream.ex:1553: Enumerable.Stream.do_each/4
    (elixir) lib/stream.ex:637: Stream.run/1
```

### Expected behavior


I expected the constraint error to be displayed.

**Comment by josevalim** [comment]
Oh, I see, it is an issue with streams rather. Did you try Ecto 3.0? We have improved this part of the code considerably, so it may be doing the correct thing now.

**Comment by samphilipd** [comment]
Thanks for the reply José. Actually this is even a problem with the following code:

```
Repo.transaction(fn ->
  query =
    from(b in Buyer,
      order_by: [asc: :id]
    )

  query
  |> Repo.all()
  |> Enum.each(fn buyer ->
    Repo.update!(Ecto.Changeset.change(buyer, %{close_io_lead_id: nil})) # Fails with not-null constraint
  end)
end)
```

So I don't think it has to do with streams.

We are not quite ready to upgrade to Ecto 3.0, will update this ticket when we do though!

**Comment by josevalim** [comment]
Oh, sweet! I will investigate, thanks!

**Comment by josevalim** [comment]
This is fixed in master in Ecto 3.0. Here is a commit that adds a test to make sure we won't have regressions: https://github.com/elixir-ecto/ecto_sql/commit/165178c64c9e59cc61176ecfe9a7a6e82e4af63d

Thanks!
